Source/Core/Common/Arm64Emitter.cpp

```diff
@@ -2761,21 +2876,117 @@ void ARM64FloatEmitter::FMUL(u8 size, ARM64Reg Rd, ARM64Reg Rn, ARM64Reg Rm, u8
 
 void ARM64FloatEmitter::ABI_PushRegisters(BitSet32 registers)
 {
-	for (auto it : registers)
-		STR(128, INDEX_PRE, (ARM64Reg)(Q0 + it), SP, -16);
+	bool bundled_loadstore = false;
+	int num_regs = registers.Count();
+
+	if (num_regs == 2)
+	{
+		int i = 0;
+		ARM64Reg regs[2];
+		for (auto it : registers)
+			regs[i++] = (ARM64Reg)(Q0 + it);
+
+		STP(128, INDEX_PRE, regs[0], regs[1], SP, -32);
+		return;
+	}
+
+	for (int i = 0; i < 32; ++i)
+	{
+		if (!registers[i])
+			continue;
 
+		int count = 0;
+		while (++count < 4 && (i + count) < 32 && registers[i + count]) {}
+		if (count > 1)
+		{
+			bundled_loadstore = true;
+			break;
+		}
+	}
+
+	if (!bundled_loadstore)
+	{
+		for (auto it : registers)
+			STR(128, INDEX_PRE, (ARM64Reg)(Q0 + it), SP, -16);
+	}
+	else
+	{
+		m_emit->SUB(SP, SP, num_regs * 16);
+		for (int i = 0; i < 32; ++i)
+		{
+			if (!registers[i])
+				continue;
+
+			int count = 0;
+
+			// 0 = true
+			// 1 < 4 && registers[i + 1] true!
+			// 2 < 4 && registers[i + 2] true!
+			// 3 < 4 && registers[i + 3] true!
+			// 4 < 4 && registers[i + 4] false!
+			while (++count < 4 && (i + count) < 32 && registers[i + count]) {}
+
+			ST1(64, count, INDEX_POST, (ARM64Reg)(Q0 + i), SP);
```

Comment:
This actually violates the AArch64 ABI (aapcs64) since data would be stored beyond the current stack pointer. Although linux normally allows for a small red zone (safe area below SP that won't be touched during exception handling) other platforms aren't so nice and it is technically a bug.